File: api/Graph.py
from Vertex import Vertex

# Based on https://pythonandr.com/2016/07/28/implementing-undirected-graphs-in-python/

class Graph:
    def __init__(self):
        self.vertices = {}
        self.vert_objs = {}
        # Specifically keep track of teams, so as to know where to pick up from next time.
        self.teams = []

    # Add a vertex to the graph, include its current neighbors into the adjacency list
    def add_vertex(self, vertex):
        if isinstance(vertex, Vertex):
            self.vertices[vertex.id] = vertex.neighbors
            self.vert_objs[vertex.id] = vertex

    # Add an edge between two vertices. If any of the vertices were not already in the graph, then add
    # them to the graph. For our purposes, the API doesn't ensure that the two vertices are of opposite
    # types (might be something to look into)
    def add_edge(self, v_from, v_to):
        if isinstance(v_from, Vertex) and isinstance(v_to, Vertex):
            # Change this so the var names aren't so long
            if v_from.id not in self.vert_objs.keys():
                self.add_vertex(v_from)
            if v_to.id not in self.vert_objs.keys():
                self.add_vertex(v_to)
            self.vert_objs[v_from.id].add_neighbor(self.vert_objs[v_to.id])
            self.vertices[v_from.id] = self.vert_objs[v_from.id].neighbors
            self.vertices[v_to.id] = self.vert_objs[v_to.id].neighbors

    # Return the adjacency list of the graph
    def adjacency_list(self):
        if len(self.vertices) >= 1:
            return [str(key) + ":" + str(self.vertices[key]) for key in self.vertices.keys()]
        else:
            return {}

    # Return all vertices in the graph by ID
    def all_vertices(self):
        return self.vertices.keys()

    # Return all player vertices
    def all_player_vertices(self):
        players = []
        for v in self.vertices:
            if v.isPlayer:
                players.append(v)
        return players

    # Return all team vertices (including yearly rosters)
    def all_team_vertices(self):
        teams = []
        for v in self.vertices:
            if not v.isPlayer:
                teams.append(v)
        return teams

    # Code from https://www.geeksforgeeks.org/building-an-undirected-graph-and-finding-shortest-path-using-dictionaries-in-python/
    # Returns an empty list if no path exists

    def find_path(self, source_id, target_id):

        # Edge case where they're the same
        if source_id == target_id:
            return [source_id]

        visited = []    # Tracks visited vertices
        queue = [[source_id]]   # Vertices to visit

        while queue:
            path = queue.pop(0)
            node = path[-1]

            # If the node was visited already, can skip
            if node not in visited:
                neighbours = self.vertices[node]

                # Loop to iterate over the neighbors of the node
                for neighbour in neighbours:
                    new_path = list(path)
                    new_path.append(neighbour)
                    queue.append(new_path)

                    # Check if the neighbor node is the goal
                    if neighbour == target_id:
                        return new_path
                visited.append(node)

        # Condition when the nodes
        # are not connected
        return []

    # Get all the teammates a player has ever been on a roster with (in essence,
    # all nodes that are two edges away (one for team, one for players))
    def get_all_teammates(self, player_id):

        pass

    # Get a string representation of the graph
    def __repr__(self):
        # return str(g.adjacencyList()) + '\n' + '\n' + str(g.adjacencyMatrix())
        return str(self.adjacency_list())

File: src/components/Graph/Graph.js
const vertices = require("./graph.json");
const i2n = require("./id_to_name.json");
const n2i = require("./name_to_id.json");
const vert_objs = require("./vert_objs.json");
const fuzzysort = require("fuzzysort");

class Graph {
  constructor(vertices, i2n, n2i, vert_objs) {
    this.vertices = vertices;
    this.id_to_name = i2n;
    this.name_to_id = n2i;
    this.vert_objs = vert_objs;
    this.names = [];
    this.players = [];
    this.size = 0;

    for (const name in n2i) {
      this.names.push(name);
    }

    this.size = this.names.length;

    for (const id in vert_objs) {
      if (vert_objs[id].isPlayer) {
        this.players.push(vert_objs[id].searchName);
      }
    }
  }

  // TESTING: Split the path function so that it doesn't block the UI, testing some techniques here
  uiBlockingTest(sourceId, targetId, graphReturnFunc, graphCallback) {
    const visited = []; // Tracks visited vertices
    const queue = [[sourceId]]; // Vertices to visit
    const verts = this.vertices;
    const maxTimePerChunk = 50; // Tested to be a pretty good result

    function now() {
      return new Date().getTime();
    }

    if (!verts[sourceId] || !verts[targetId]) {
      graphReturnFunc([]);
      return;
    }

    // Edge case where they're the same
    if (sourceId === targetId) {
      graphReturnFunc([sourceId]);
      return;
    }

    function doChunk() {
      const startTime = now();
      while (queue.length > 0 && now() - startTime <= maxTimePerChunk) {
        let path = queue.shift();
        let node = path[path.length - 1];
        if (!visited.includes(node)) {
          let neighbors = verts[node];
          // Loop to iterate over the neighbors of the node
          for (const neighbor of neighbors) {
            let newPath = path.slice();
            newPath.push(neighbor);
            queue.push(newPath);
            // Check if the neighbor node is the goal
            if (neighbor === targetId) {
              graphReturnFunc(newPath);
              return;
            }
          }
          visited.push(node);
        }
      }
      if (queue.length > 0) {
        const mounted = graphCallback(visited.length);
        if (!mounted) {
          return;
        }
        setTimeout(doChunk, 0);
      } else {
        graphReturnFunc([]);
        return;
      }
    }
    doChunk();
  }

  // Return a random item (either name or player)
  randomName(category) {
    if (category === "players") {
      return this.players[Math.floor(Math.random() * this.players.length)];
    } else if (category === "teams") {
      return "Not yet implemented";
    } else if (category === "all") {
      return this.names[Math.floor(Math.random() * this.names.length)];
    } else {
      return "Error";
    }
  }

  // Returns the closest matching names of all
  bestNames(input, coll, amount = 20) {
    let collection;
    switch (coll) {
      case "all":
        collection = this.names;
        break;
      case "players":
        collection = this.players;
        break;
      case "teams":
        collection = this.teams;
        break;
      default:
        collection = [];
    }

    const inColl = collection.includes(input);

    const res = fuzzysort.go(input, collection, {
      limit: amount,
      threshold: inColl ? -100 : -Infinity,
      allowTypo: true,
    });
    const arr = [];
    for (const r of res) {
      arr.push(r.target);
    }
    return arr;
  }

  getTeammates(sourceId) {
    // Handle case where either of them doesn't exist in the graph
    if (!sourceId) {
      return [];
    }

    const teammates = [];
    const visited = []; // Tracks visited vertices
    const queue = [[sourceId]]; // Vertices to visit

    while (queue.length > 0) {
      let path = queue.shift();
      let node = path[path.length - 1];

      // If the node was visited already, can skip
      // Might not need this for the teammate one
      if (!visited.includes(node)) {
        let neighbors = this.vertices[node];

        // Case where two nodes are already in the path, the original player and the team. Means the third node
        // about to be added must be a teammate
        if (path.length === 2) {
          for (const neighbor of neighbors) {
            if (!teammates.includes(neighbor)) {
              teammates.push(neighbor);
            }
          }
        } else {
          // Loop to iterate over the neighbors of the node
          for (const neighbor of neighbors) {
            let newPath = path.slice();
            newPath.push(neighbor);
            queue.push(newPath);
          }
          visited.push(node);
        }
      }
    }

    // Remove the sourceId from the teammates
    const idx = teammates.indexOf(sourceId);
    if (idx > -1) {
      teammates.splice(idx, 1);
    }
    return teammates;
  }

  path(sourceId, targetId) {
    // Handle case where either of them doesn't exist in the graph
    if (
      !sourceId ||
      !targetId ||
      !vertices.includes(sourceId) ||
      !vertices.includes(targetId)
    ) {
      return [];
    }

    // Edge case where they're the same
    if (sourceId === targetId) {
      return [sourceId];
    }

    const visited = []; // Tracks visited vertices
    const queue = [[sourceId]]; // Vertices to visit

    while (queue.length > 0) {
      let path = queue.shift();
      let node = path[path.length - 1];

      // If the node was visited already, can skip
      if (!visited.includes(node)) {
        let neighbors = this.vertices[node];

        // Loop to iterate over the neighbors of the node
        for (const neighbor of neighbors) {
          let newPath = path.slice();
          newPath.push(neighbor);
          queue.push(newPath);

          // Check if the neighbor node is the goal
          if (neighbor === targetId) {
            return newPath;
          }
        }
        visited.push(node);
      }
    }
    // Condition when the nodes are not connected
    return [];
  }
}

const g = new Graph(vertices, i2n, n2i, vert_objs);

export default g;
